src/AgentHeaderDescriptor.php

```diff
@@ -43,44 +41,87 @@ class AgentHeaderDescriptor
     // TODO(michaelbausor): include bumping this version in a streamlined
     // release process. Issue: https://github.com/googleapis/gax-php/issues/48
     const GAX_VERSION = '0.7.0';
+    const UNKNOWN_HEADER = '';
 
-    private $clientName;
-    private $clientVersion;
-    private $codeGenName;
-    private $codeGenVersion;
-    private $gaxVersion;
-    private $phpVersion;
+    private $metricsHeaders;
 
     /**
      * @param array $headerInfo {
-     *     Required.
+     *     Optional.
      *
-     *     @type string $clientName the name of the client application.
-     *     @type string $clientVersion the version of the client application.
+     *     @type string $phpVersion the PHP version.
+     *     @type string $libName the name of the client application.
+     *     @type string $libVersion the version of the client application.
      *     @type string $codeGenName the code generator name of the client library.
      *     @type string $codeGenVersion the code generator version of the client library.
      *     @type string $gaxVersion the GAX version.
-     *     @type string $phpVersion the PHP version.
+     *     @type string $grpcVersion the gRPC version.
+     *     @type string $additionalMetrics a dictionary of additional metrics for the header.
      * }
      */
     public function __construct($headerInfo)
     {
-        $this->clientName = $headerInfo['clientName'];
-        $this->clientVersion = $headerInfo['clientVersion'];
-        $this->codeGenName = $headerInfo['codeGenName'];
-        $this->codeGenVersion = $headerInfo['codeGenVersion'];
-        $this->gaxVersion = $headerInfo['gaxVersion'];
-        $this->phpVersion = $headerInfo['phpVersion'];
+        $metricsHeaders = [];
+
+        // The ordering of the headers is important. We use the fact that $metricsHeaders is an
+        // ordered dict. The desired ordering is:
+        //      - phpVersion (gl-php/)
+        //      - clientName (e.g. gccl/)
+        //      - codeGenName (e.g. gapic/)
+        //      - gaxVersion (gax/)
+        //      - grpcVersion (grpc/)
+        //      - any additional metrics
+
+        $phpVersion = isset($headerInfo['phpVersion'])
+            ? $headerInfo['phpVersion']
+            : phpversion();
+        $metricsHeaders['gl-php'] = $phpVersion;
+
+        if (isset($headerInfo['libName'])) {
+            $clientVersion = isset($headerInfo['libVersion'])
+                ? $headerInfo['libVersion']
+                : AgentHeaderDescriptor::UNKNOWN_HEADER;
+            $metricsHeaders[$headerInfo['libName']] = $clientVersion;
+        }
+
+        if (isset($headerInfo['codeGenName'])) {
```

Comment:
This should always be set, and should always be `gapic`. We're not worried about the use case of anyone else using GAX directly.